### PR TITLE
Use ClassCellNode for __class__ so that it's available in methods

### LINKED ISSUE #2912: Cython mishandles PEP-3135 __class__ cell in methods
Hello, I am having trouble with some inconsistent behaviour round the value of `__class__` in Cythonized code. This is on Python **3.6.7** with Cython **0.29.1**. 

Briefly, in uncythonized code `__class__` inside a method on a `class` refers to the `class` object, whereas once cythonized, it seems to refer to `<class 'module'>`. This behaviour is a bit unintuitive, and has caused a few issues in a codebase I'm working on, but it's possible I have misunderstood the language semantics! 

I have tried to search for reasons / documentation why this is the case, but I couldn't turn up anything. I apologise if there is something really obvious I'm missing here; let me know if so and I'll close the issue at once. 

Example:

`minimal.py`:
```
class Something:
    def method(self):
        print(__class__)
```

(in an interpreter, uncythonized):
```
from minimal import Something
Something().method()
# prints <class 'minimal.Something'>
```
(after running `cythonize minimal.py -3`):
```
from minimal import Something
Something().method()
# prints <class 'module'>   
```

I believe there is an obvious workaround -- in instance methods, `self.__class__` should still have the correct value. But I was curious if this is intended behaviour, and if so, is there something in the docs about it?

### PR DESCRIPTION
**Pull request by marginalhours**
Closes #2912 

Per the comment on the issue, this turns `NameNodes` for `__class__` into `ClassCellNodes`, so that you can call `__class__` inside methods without qualifying with `self`. I believe that this fix is effectively just replacing `__class__` with `<name of class>` in these expressions, EG, no `__class__` local ends up created. 

This is quite a minimal fix. There are a few remaining differences in `__class__` behaviour but I wasn't sure how to proceed with them, so I thought I'd raise a PR for comment. 

Basically, in non-Cythonized Python it appears that `__class__` is a bit of a "magic" local. If it appears anywhere in a method body, then `locals()` will contain `__class__` pointing to the containing class, anywhere in the function, but if it doesn't, then it won't. So:

```
class A:
    def b(self):
        print(locals())
        __class__
```
`A().b()` will print `['self', '__class__']`. As would 

```
class A:
    def b(self):
        __class__
        print(locals())
```

But:

```
class A:
    def b(self):
        print(locals())
```
would print only `['self']`

This behaviour is hard to achieve with the current PR, because instead of creating a local called `__class__` we are just replacing the `NameNode` with a `ClassCellNode`. As a result, when cythonized, in all the above examples `locals()` just prints `['self']`. 

Off the top of my head, some approaches that might be better: 

- Instead of replacing the `NameNode`, inject an `entry` into the current scope called `__class__` pointing to the containing `__class__` object. This requires rewriting `FuncLocalsExprNode` so that it iterates over the keys in `entries` rather than calling `.name` on each one. I tried this and it sort of worked but it was a bit messy manually updating the entries dictionary, maybe there's a better way?

- From a higher level in the tree (the `DefNode`?) iterating through all of the statements in the body checking to see if any of them contain `__class__` and, if so, injecting it as a local.  

- Without bothering to see if it's in the body or not, just always injecting `__class__` into the method scope (it's harmless if unused and I don't think it's unreasonable to interpret PEP3135 as implying it should always be in the `locals()` symbol table. However, I am far from an expert and there is potentially a good reason why Python doesn't put it into `locals()` for methods not containing `__class__`.)

Eager to get any and all feedback on the above! This is my first time touching the Cython codebase so I might have missed some things, but it's been really fun to dig into.

### CODE REVIEW

File: Cython/Compiler/ParseTreeTransforms.py [hunk 1]
```diff
@@ -3114,6 +3115,21 @@ def _inject_eval(self, node, func_name):
                     node.pos, self.current_scope_node(), lenv))
         return node
 
+    def _inject_class(self, node):
+        if node.name == u'__class__':
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: Since this covers the complete function body (and the function name suggests that it does what it says), I'd move this `if` into `visit_NameNode()` above.

File: Cython/Compiler/ParseTreeTransforms.py [hunk 2]
```diff
@@ -3114,6 +3115,21 @@ def _inject_eval(self, node, func_name):
...
+            # bare __class__ reference inside function
+            def_node = self.current_scope_node()
+            if (not isinstance(def_node, Nodes.DefNode) or not def_node.args or
+                len(self.env_stack) < 2):
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: Max line length in Cython is 120, so no need to wrap this line.

File: Cython/Compiler/ParseTreeTransforms.py [hunk 3]
```diff
@@ -3114,6 +3115,21 @@ def _inject_eval(self, node, func_name):
...
+                return node
+            class_node, class_scope = self.env_stack[-2]
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: This is a bit of an assumption. :)
Instead, you can follow the `parent_scope` chain of the current scope to find the outer class.

**Comment by scoder** [comment]
There is a dedicated `locals()` implementation in Cython, which literally just builds a dict of local variables and their current values. I'd suggest to simply implement `__class__` in there as a special case variable that appears when the `DefNode` uses it. See `_inject_locals()` in `ParseTreeTransforms.py`.

File: Cython/Compiler/ParseTreeTransforms.py (hunk 3 above)
Comment (scoder) [reply Cython/Compiler/ParseTreeTransforms.py]: I just realised that the current `super()` implementation uses exactly this code. Probably worth at least extracting it into a common method to make it easier to change later.

**Comment by MisterKeefe** [comment]
I've modified this to extract some common functionality from `_inject_super` and `_inject_class` as helper methods. Also, now instead of replacing the `NameNode` with a `ClassCellNode`, if there's a `NameNode` called `__class__` we just inject `__class__` into the local environment for the method.

File: Cython/Compiler/ParseTreeTransforms.py [hunk 4]
```diff
@@ -3114,17 +3116,41 @@ def _inject_eval(self, node, func_name):
                     node.pos, self.current_scope_node(), lenv))
         return node
 
+    def _check_inside_class(self, def_node):
+        return isinstance(def_node, Nodes.DefNode) and def_node.args and len(self.env_stack) >= 2
+        
+
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: Superfluous blank line.

File: Cython/Compiler/ExprNodes.py [hunk 5]
```diff
@@ -9970,8 +9970,7 @@ def analyse_types(self, env):
 
 class FuncLocalsExprNode(DictNode):
     def __init__(self, pos, env):
-        local_vars = sorted([
-            entry.name for entry in env.entries.values() if entry.name])
+        local_vars = sorted([key for key, entry in env.entries.items() if entry.name])
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: Why was this change needed?

File: Cython/Compiler/ParseTreeTransforms.py [hunk 6]
```diff
@@ -3114,17 +3116,41 @@ def _inject_eval(self, node, func_name):
...
+        def_node = self.current_scope_node()
+        if not self._check_inside_class(def_node):
+            return
+        
+        class_node, class_scope = self._get_current_class_and_scope()
+        if class_scope.is_py_class_scope:
+            def_node.requires_classobj = True
+            class_node.class_cell.is_active = True
+            lenv.entries[EncodedString(u"__class__")] = class_node.target.entry
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: It shouldn't be necessary to use an `EncodedString` here – unless you use the keys for building the locals dict, as your change does above. Or did you stumble over other things that broke along the way?

**Comment by scoder** [comment]
(BTW, I fixed the travis test problems, merging master should resolve them for this branch.)

File: Cython/Compiler/ExprNodes.py (hunk 5 above)
Comment (MisterKeefe) [reply Cython/Compiler/ExprNodes.py]: The way I am injecting the `__class__` entry into the scope (for some class `C`), it looks like: `{ "__class__": Entry(<some memory address>, name=C, type=Python object)}`. Without this change, this results in the call to `locals()` returning a `dict` with two keys: `self` and `C`. With this change, it results in the two keys `self` and `__class__`. 

I'm not sure I fully understand Entries. Would it be better to create a new one which points to the same place as `C` but has `name=__class__`?

File: Cython/Compiler/ParseTreeTransforms.py (hunk 6 above)
Comment (MisterKeefe) [reply Cython/Compiler/ParseTreeTransforms.py]: See above comment -- I think the alternative would be to create an `Entry` with `name=__class__`. I'll try to do it that way.

Comment (scoder) [reply Cython/Compiler/ParseTreeTransforms.py]: Yes, there should be a new `Entry`. You can use `env.declare_var(name, type, pos)` to create one, like `NameNode.analyse_target_declaration()` does.

Comment (MisterKeefe) [reply Cython/Compiler/ParseTreeTransforms.py]: OK, I've got as far as creating a new `Entry` for `__class__`, but then it fails with local variable `__class__` referenced before assignment. I can also set the name to point to `class_node.target.entry.name`, but that still fails with a similar error. I will keep digging - there doesn't seem to be an obvious way to make these entries synonymous.

**Comment by scoder** [comment]
> it fails with local variable __class__ referenced before assignment

Lovely. What strikes here is the control flow analysis in `FlowControl.py`, which correctly detects an unbound local variable.

Here's something I tried in Py3.6:
```
>>> class C:
...   def meth(self, x):
...     if x: __class__ = x
...     print(__class__)
... 
>>> C().meth(1)
1
>>> C().meth(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in meth
UnboundLocalError: local variable '__class__' referenced before assignment
```
This suggests that there is no assignment to `__class__` happening if it's already defined, so we need to take care that we don't just declare it if it's already there, or unconditionally fake an assignment to it. That test should be added, BTW. Another test to add is this:
```
>>> class C:
...    def meth(self):
...       def inner(): return __class__
...       return inner
>>> C().meth()()
<class '__main__.C'>
```
suggesting that the name must exist as an actual Entry to make it available in closures. Finally, declaring a class inside of a closure function (or, even better, inside of a method of another class) is also worth testing.

What about this: we interpret `__class__` not as a local variable but as a pure reference to the class itself. Meaning, we use the original class `Entry` for the `NameNode`, but do not put it into the local variable entries (because it's not a local variable entry). We then implement something in `locals()` that explicitly injects it there if the def-node needs it but the local scope does not provide it.

Could you try that?

**Comment by scoder** [comment]
Hmm, hang on, there's another test case that might fail here:
```
class C(object):
    def meth(): return __class__
OldC = C
C = None
print(OldC().meth())
```
I guess I have to think this through a bit more…

**Comment by MisterKeefe** [comment]
I've included a few more test cases for some nested and edge cases. The approach is what you recommended: 

(1) Inject a `NameNode` for `__class__` references 
(2) Add an attribute `has_class_reference` to `DefNode`s to keep track of which ones refer to `__class__`
(3) In `_inject_locals()`, if the relevant `DefNode`'s `has_class_reference` attribute is `True`, inject an `Entry` called `__class__` pointing at the relevant `ClassNode`.

File: Cython/Compiler/ParseTreeTransforms.py [hunk 7]
```diff
@@ -3067,7 +3080,7 @@ def _inject_locals(self, node, func_name):
                           % len(node.args))
                 if len(node.args) > 0:
                     return node # nothing to do
-            return ExprNodes.LocalsExprNode(pos, self.current_scope_node(), lenv)
+            return ExprNodes.LocalsExprNode(pos, def_node, lenv)
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: I think `def_node` might not have been assigned here in some cases. Feel free to always assign it outside of the condition above.

File: tests/run/pep3135_class_cell.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                def inner(self):
+                    return __class__
+            return J().inner
+        return inner
+
+
+class K:
+    """
+    >>> OldK = K
+    >>> K = None
```
Comment (scoder) [inline tests/run/pep3135_class_cell.py]: I don't think that this really changes the global name *in the module*, where the method looks it up. Please move these assignments into the main module code, just after defining the class `K`.

File: tests/run/py3k_super.pyx [hunk 9]
```diff
@@ -45,6 +45,111 @@ class B(A):
...
 
+class C(A):
+    """
+    >>> obj = C()
+    >>> obj.method_1()
+    2
+    >>> obj.method_2()
+    3
+    >>> obj.method_3()
+    ['__class__', 'self']
```
Comment (scoder) [inline tests/run/py3k_super.pyx]: And is `__class__` really `C` here (or `A` by accident)?

File: Cython/Compiler/ParseTreeTransforms.py [hunk 10]
```diff
@@ -3114,17 +3127,42 @@ def _inject_eval(self, node, func_name):
...
     def _inject_super(self, node, func_name):
         lenv = self.current_env()
         entry = lenv.lookup_here(func_name)
         if entry or node.args:
             return node
         # Inject no-args super
         def_node = self.current_scope_node()
-        if (not isinstance(def_node, Nodes.DefNode) or not def_node.args or
-            len(self.env_stack) < 2):
+        if not self._check_inside_class(def_node) or not def_node.args:
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: I think it's also worth testing if a method `meth(*args)` satisfies this condition. IIRC, `*args` and `**kwargs` are not part of the `.args` list of the def-node.

File: Cython/Compiler/ParseTreeTransforms.py [hunk 11]
```diff
@@ -3114,17 +3127,42 @@ def _inject_eval(self, node, func_name):
...
         return node
 
+    def _check_inside_class(self, def_node):
+        return isinstance(def_node, Nodes.DefNode) and any(isinstance(node, Nodes.ClassDefNode) for node, _ in self.env_stack)
+
+    def _get_current_class_and_scope(self):
+        return next(
+            (node, scope)
+            for node, scope
+            in self.env_stack[::-1]
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: IMHO, `for` and `in` always belong into the same line, especially if an `if` follows, with which `in` is easily confused visually.

File: Cython/Compiler/Nodes.py [hunk 12]
```diff
@@ -2787,6 +2787,7 @@ class DefNode(FuncDefNode):
 
     is_staticmethod = False
     is_classmethod = False
+    has_class_reference = False
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: :+1:

**Comment by scoder** [comment]
You could also try injecting the `__class__` only in `LocalsExprNode`, but that might break closures.

File: Cython/Compiler/ParseTreeTransforms.py (hunk 10 above)
Comment (MisterKeefe) [reply Cython/Compiler/ParseTreeTransforms.py]: I might have gone a bit overboard with this, but there's new tests in `py3k_super.pyx` for `@staticmethods` and `@classmethods`.

File: tests/run/py3k_super.pyx (hunk 9 above)
Comment (MisterKeefe) [reply tests/run/py3k_super.pyx]: Good point; I've added tests for this below.

File: tests/run/pep3135_class_cell.py (hunk 8 above)
Comment (scoder) [reply tests/run/pep3135_class_cell.py]: Ok, what I meant was that this should work:
```
class K(object):
    """
    >>> OldK().method().__name__
    'K'
    """
    def method(self): return __class__

OldK = K
K = None
```
It probably doesn't currently.

Comment (MisterKeefe) [reply tests/run/pep3135_class_cell.py]: Yeah, that currently breaks with the predictable `'NoneType' object has no attribute '__name__'`.

Comment (MisterKeefe) [reply tests/run/pep3135_class_cell.py]: So what's happening here is that `K` is being rebound to `None` in the outer scope, but the `DefNode`'s scope should hold onto the reference, if I have that right.

**Comment by MisterKeefe** [comment]
> Using the `NameNode` to refer to the class does not work since it might be overwritten, e.g. via class decorators or simply name assignments.

Yes, that makes sense. I've added a couple of test cases for this and updated to use an `AttributeNode` pointing to `self.__class__`, which I think gets past that problem.

**Comment by MisterKeefe** [comment]
Another strange test case: 

`class R(object):
    def method(self):
        return 1

class S(R):
    """
    >>> S().method()
    ...
    RuntimeError: super(): __class__ cell not found
    """
    def method(self):
        __class__ = None
        return super().method()
`

File: tests/run/pep3135_class_cell.py [hunk 13]
```diff
@@ -117,3 +117,31 @@ def method(self): return __class__
     'K'
     """
 }
+
+class L:
+    """
+    >>> OldL().method().__name__
+    'L'
+    """
+    def method(self): return self.__class__
```
Comment (scoder) [inline tests/run/pep3135_class_cell.py]: I think this should read `return __class__` and not `return self.__class__`.

File: Cython/Compiler/ParseTreeTransforms.py [hunk 14]
```diff
@@ -3153,7 +3153,8 @@ def _inject_class(self, node):
 
         if class_scope.is_py_class_scope:
             def_node.has_class_reference = True
-            return ExprNodes.NameNode(node.pos, name=class_node.target.name)
+            self_node = ExprNodes.NameNode(pos=node.pos, name=u'self')
+            return ExprNodes.AttributeNode(pos=node.pos, obj=self_node, attribute=EncodedString(u'__class__'))
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: No, `self.__class__` is not correct since it's not compile time constant. This should refer to the class that defines the method, not the one on which the method is called. The reference should come from `ClassCellNode` (or something similar).

**Comment by mattip** [comment]
Since gh-2912 is marked as blocking it would be good to finish this. @MisterKeefe would you like some help?

**Comment by MisterKeefe** [comment]
@mattip Yes, thank you -- I kind of stalled out on this one, sorry. It might be worth reverting to `1ede6da` on this branch as a starting-point; while that didn't have perfect semantics for __class__ it covered several of the test cases. 

Where I got to with thinking about it was that for classes with an internal `__class__` reference, there should be two variable declarations in the generated code with the second pointing to the first. EG for this case: 

```
class A:
    def method():
        print(__class__)

class B(A):
    pass

A = None

B().method()
```

Resolving `__class__` as the same `NameNode` as `A` doesn't work because that disappears after `A` is assigned to `None.` But if a duplicate declaration was emitted for `__class__` which didn't get removed by the `None`, that might cover this case.

**Comment by da-woods** [comment]
This may be stating what's already known, but Python handles this by putting the class in the `__closure__`

    class A:
       def f(self):
          return __class__
    print(A.f.__closure__)  # prints (<cell at 0x7faba5bd31c0: type object at 0x564d21dedaa0>,)

I think Cython basically has to do the same thing (since as you point out, any attempt to look it up at runtime can end up with stuff overridden).

It's possible that this will turn out to be more complicated than I'm imagining (because class-scopes aren't normally looked at for closures), but the first thing I'd try is to set up the entry for `__class__` with `.from_closure` (or similar) and see if it can resolve to the correct thing

**Comment by da-woods** [comment]
> This may be stating what's already known, but Python handles this by putting the class in the `__closure__`
> 
> ```
> class A:
>    def f(self):
>       return __class__
> print(A.f.__closure__)  # prints (<cell at 0x7faba5bd31c0: type object at 0x564d21dedaa0>,)
> ```
> 
> I think Cython basically has to do the same thing (since as you point out, any attempt to look it up at runtime can end up with stuff overridden).
> 
> It's possible that this will turn out to be more complicated than I'm imagining (because class-scopes aren't normally looked at for closures), but the first thing I'd try is to set up the entry for `__class__` with `.from_closure` (or similar) and see if it can resolve to the correct thing

Looking at this again, it's `ClassCellNode` (as suggested) does something very similar although using a C-level mechanism, so that's the better advice.

**Comment by da-woods** [comment]
I've tried to finish this is https://github.com/cython/cython/pull/4094. It turns out to be a little more complicated than it looked.
